components/constellation/constellation.rs

```diff
@@ -1376,6 +1384,94 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
         }
     }
 
+    fn handle_manual_redirect(&mut self, id: PipelineId, load_data: LoadData) {
+        let mut req_init = RequestInit {
+            url: load_data.url.clone(),
+            method: load_data.method,
+            destination: Destination::Document,
+            credentials_mode: CredentialsMode::Include,
+            use_url_credentials: true,
+            origin: load_data.url.clone(),
+            pipeline_id: Some(id),
+            referrer_url: load_data.referrer_url,
+            referrer_policy: load_data.referrer_policy,
+            headers: load_data.headers,
+            body: load_data.data,
+            redirect_mode: RedirectMode::Manual,
+            .. RequestInit::default()
+        };
+
+        let res_init = ResponseInit {
+            url: load_data.url.clone(),
+            .. ResponseInit::default()
+        };
+
+        // keep looping until we get a response that doesn't redirect
+        loop {
+            let (sender, receiver) = ipc::channel().expect("Failed to create IPC channel!");
+            if let Err(e) = self.public_resource_threads.send(
+                                CoreResourceMsg::FetchRedirect(req_init.clone(),
+                                res_init.clone(),
+                                sender)) {
+                warn!("Exit resource thread failed ({})", e);
+            }
+            match receiver.recv() {
+                Ok(recvd) => {
+                    let data = match recvd {
+                        FetchResponseMsg::ProcessResponse(res_metadata) => {
+                            match res_metadata {
+                                Ok(metadata) => {
+                                    match metadata {
+                                        FetchMetadata::Filtered {
+                                            filtered,
+                                            unsafe_: unsafe_metadata
+                                        } => Some((filtered, unsafe_metadata)),
+                                        _ => None
+                                    }
+                                },
+                                Err(_) => None,
+                            }
+                        },
+                        _ => break
+                    };
+                    match data {
+                        Some((FilteredMetadata::OpaqueRedirect, _)) => {
+                            if let Some((_, unsafe_metadata)) = data {
+                                match unsafe_metadata.headers {
+                                    Some(headers) => match headers.get::<Location>() {
+                                        Some(url) => {
+                                            match ServoUrl::parse(url) {
```

Comment:
Although it occurs to me that that would break the following case:
* fetch http://foo.com
* redirect to http://bar.com
* redirect to /foo.html
However, the response contains a [url list](https://fetch.spec.whatwg.org/#concept-response-url-list), so it should be possible to parse the location value using the most recent entry in the url list, and progressively move backwards until a successfull URL is parsed.